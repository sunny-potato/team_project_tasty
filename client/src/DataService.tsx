import axios from 'axios';
import React, { useEffect, useState } from 'react';
axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type RecipeInfo = {
  id: number;
  name: string;
  meal_type: string;
  new: boolean;
  popular: boolean;
  description: string;
};

export type Ingredient = {
  ingredients_id: number;
  ingredient: string;
  amount: number;
  unit_id: number;
  unit: string;
};

export type Recipe = {
  recipeInfo: RecipeInfo;
  ingredients: Ingredient[];
};
export type Unit = {
  id: number;
  unit: string;
};
export type UnitsList = {
  units: Unit[];
};

export type EachIngredient = {
  id: number;
  ingredient: string;
};

export type IngredientsList = {
  ingredients: EachIngredient[];
};

class DataService {
  /* Get a specific recipe with known id */

  get(id: number | string) {
    // Use this for testing locally only
    // return new Promise<Recipe>((resolve, reject) => {
    //   resolve({
    //     recipeInfo: {
    //       id: 1,
    //       name: 'The best pasta in the world',
    //       meal_type: 'Vegan',
    //       new: true,
    //       popular: true,
    //       description:
    //         'Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolore, corporis. Distinctio deserunt aperiam hic voluptatum, ut a molestias cumque tenetur consequatur, ad dolores voluptatem provident eveniet, porro magnam est veritatis.',
    //     },
    //     ingredients: [
    //       { name: 'Olive oil', amount: 3, unit: 'tbsp' },
    //       { name: 'Onion', amount: 1, unit: 'pcs' },
    //       { name: 'Garlic', amount: 3, unit: 'cloves' },
    //       { name: 'Chopped Tomato', amount: 400, unit: 'g' },
    //       { name: 'Dried spaghetti', amount: 300, unit: 'g' },
    //     ],
    //   });
    // });

    return axios
      .get<Recipe>('/recipe/' + id)
      .then((response) => response.data)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }
  // Get an array of all recipes

  getAll() {
    return axios
      .get<[]>('/recipe/')
      .then((response) => response.data)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }
  /* Post new recepie */
  create(data: Recipe) {
    return axios
      .post<{ id: number }>('/recipe', { data })
      .then((response) => response.data.id)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }
  /* Edit an existing recipe by known id */

  edit(id: number) {
    return alert('function not implementet yet');

    /* Delete a recipe known id */
  }
  delete(id: number) {
    return axios
      .delete('/recipe/' + id)
      .then((response) => response.data)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }
  /* Get all ingredients */
  getAllIngredients() {
    return axios
      .get('/ingredient')
      .then((response) => response.data)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }

  /* Get all units */
  getAllUnits() {
    return axios
      .get<[]>('/unit')
      .then((response) => response.data)
      .catch((error) => {
        console.error(error);
        throw error;
      });
  }
}

const dataService = new DataService();
export default dataService;
