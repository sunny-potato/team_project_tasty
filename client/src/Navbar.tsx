import React from 'react';
import { Link } from 'react-router-dom';
import './index.css';

export function Navbar() {
  return (
    <nav className="NavMain">
      <Link to="/" className="Logo_title">
        Tasty
      </Link>
      <ul className="NavList">
        <li>
          <Link to="/" className="Link">
            Home
          </Link>
        </li>
        <li>
          <Link to="explore" className="Link">
            Explore
          </Link>
        </li>
        <li>
          <Link to="new" className="Link">
            Create new
          </Link>
        </li>
        <li>
          <Link to="cart" className="Link">
            Shopping list
          </Link>
        </li>
      </ul>
    </nav>
  );
}
