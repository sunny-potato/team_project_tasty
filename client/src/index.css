@import url('https://fonts.googleapis.com/css2?family=Prompt:wght@200;600&display=swap');

* {
  box-sizing: border-box;
}

body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu',
    'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin: 0;
  padding: 0;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New', monospace;
}

.NavMain {
  background-color: rgb(41, 211, 118);
  color: black;
  display: flex;
  justify-content: space-between;
  align-items: stretch;
  gap: 3rem;

  padding: 0 2rem;
}
.NavList {
  padding: 0;
  margin: 0;
  list-style: none;
  display: flex;
  gap: 2rem;
  text-decoration: none;
}
.NavList li:active {
  background-color: rgb(241, 250, 245);
}

.NavList li:hover {
  background-color: rgb(3, 190, 88);
}

.Link {
  font-family: 'Prompt', sans-serif;
  font-weight: 200;
  height: 100%;
  text-decoration: none;
  display: flex;
  align-items: center;
  padding: 0.5rem;
}

.Logo_title {
  font-family: 'Prompt', sans-serif;
  font-weight: 600;
  text-decoration: none;
  font-size: 2.5rem;
}
