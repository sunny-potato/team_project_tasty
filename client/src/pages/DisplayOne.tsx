import './PageStyling.css';
import React, { useEffect, useState, useRef, Component } from 'react';
import dataService, { Recipe, Ingredient, RecipeInfo } from '../DataService';
import { Link, useParams, useLocation } from 'react-router-dom';

export function DisplayOne() {
  const [recipe, setRecipe] = useState<Recipe>();
  const [ingredients, setIngredients] = useState<Ingredient[]>();

  //To get navigation by page and id
  const params = useParams();
  const id: string = params.id;

  // Get inital value in load, uses recipe id as source
  // Also loads values recipe and ingredients by setState

  useEffect(() => {
    let id: string = params.id;

    dataService.get(id).then((data) => {
      setRecipe(data);
      setIngredients(data.ingredients);
    });
  }, []);

  // Updates ingredients according to value chosen by user

  function changePortions(portions: number) {
    let newAmounts: Ingredient[];
    let defaultAmounts: Ingredient[] = recipe!.ingredients;

    newAmounts = defaultAmounts?.map((i) => ({
      ...i,
      amount: parseFloat(((i.amount / 4) * portions).toPrecision(2)),
    }));
    setIngredients(newAmounts);
  }

  return (
    <div className="Content-main">
      <h1 className="Recipe-title">{recipe?.recipeInfo.name}</h1>
      <h5 className="Recipe-tags">Tags: {recipe?.recipeInfo.meal_type}</h5>
      <h6 className="Recipe-new">{recipe?.recipeInfo.new ? 'This is a new recipe!' : null}</h6>
      <h6 className="Recipe-popular">
        {recipe?.recipeInfo.popular ? 'This item is popular right now!' : null}
      </h6>
      <p className="Recipe-description>">{recipe?.recipeInfo.description}</p>
      <div>
        <div>
          <label>Number of portions:&nbsp;</label>
          <select
            name="portions"
            defaultValue={4}
            onChange={(e) => changePortions(parseInt(e.currentTarget.value))}
          >
            <option value={1}>1</option>
            <option value={2}>2</option>
            <option value={3}>3</option>
            <option value={4}>4</option>
            <option value={5}>5</option>
            <option value={6}>6</option>
            <option value={7}>7</option>
            <option value={8}>8</option>
            <option value={9}>9</option>
            <option value={10}>10</option>
          </select>
        </div>
        <table className="Ingredients-main">
          <tbody>
            <tr>
              <th className="Ingredients-header">Ingredients:</th>
              <th></th>
            </tr>
            {ingredients?.map((content) => (
              <tr className="Ingredients-row" key={content.ingredients_id}>
                <td className="Ingredients-cell">
                  {content.amount}&nbsp;{content.unit}
                </td>

                <td className="Ingredients-name">{content.ingredient}</td>
              </tr>
            ))}
          </tbody>
        </table>
        <Link to="/">
          <button className="Button-navigation">Back</button>
        </Link>
        <Link to="/edit">
          <button className="Button-navigation">Edit recipe</button>
        </Link>
        <button
          className="Button-navigation"
          onClick={() => alert('Here will the like function be, not implemented yet')}
        >
          Like
        </button>
      </div>
    </div>
  );
}
