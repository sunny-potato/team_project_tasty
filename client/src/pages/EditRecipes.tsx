import React from 'react';
import { useParams } from 'react-router-dom';

export function EditRecipes() {
  return (
    <div>
      <h1>Site to edit the recipes</h1>
    </div>
  );
}
