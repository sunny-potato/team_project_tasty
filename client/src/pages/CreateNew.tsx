import React, { useState } from 'react';
import { RecipeInfo, Ingredient } from '../DataService';
import InputRecipeInfo from '../components/InputRecipeInfo';
import InputIngredients from '../components/InputIngredients';
import { useNavigate } from 'react-router-dom';
import { ValidateData, SaveEditedData } from '../components/SaveInputData';

const initialRecipeInfo: RecipeInfo = {
  id: undefined!,
  name: '',
  meal_type: '',
  new: false,
  popular: false,
  description: '',
};

const initialIngredient: Ingredient = {
  ingredients_id: undefined!,
  ingredient: '',
  amount: null,
  unit_id: 0,
  unit: 'initalUnit',
};

export function CreateNew() {
  const navigate = useNavigate();
  const [recipeInfo, setRecipeInfo] = useState<RecipeInfo>(initialRecipeInfo);
  const [ingredients, setIngredients] = useState<Ingredient[]>([initialIngredient]);
  const [isLoading, setIsLoading] = useState<boolean>(false);

  // console.log(recipeInfo);
  // console.log(ingredients);
  const onSubmit = (event: React.FormEvent) => {
    event.preventDefault();

    if (!isLoading) {
      setIsLoading(true);
      const isValid = ValidateData({ ingredients }); // current condition is always true
      if (isValid) {
        SaveEditedData({ ingredients, recipeInfo });
      }
    } else {
      alert('something wrong');
      //*********************//
      //here will be added f meal_type or description is required
    }
    setIsLoading(false);
  };

  return (
    <div>
      <h1 className="Page-title">Create new recipe</h1>
      <div className="Content-main">
        <form onSubmit={onSubmit}>
          <InputRecipeInfo recipeInfo={recipeInfo} setRecipeInfo={setRecipeInfo} />
          <InputIngredients ingredients={ingredients} setIngredients={setIngredients} />
          <div className="btn-group">
            <button type="submit" style={{ opacity: isLoading ? 0.3 : 1 }}>
              {isLoading ? 'Saving' : 'Save'}
            </button>
            <button type="button" className="cancelButton" onClick={() => navigate('/')}>
              Cancel
            </button>
          </div>
        </form>
      </div>
    </div>
  );
}
