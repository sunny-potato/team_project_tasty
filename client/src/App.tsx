import React from 'react';
import { Route, Routes } from 'react-router-dom';
import { CreateNew } from './pages/CreateNew';
import { DisplayOne } from './pages/DisplayOne';
import { EditRecipes } from './pages/EditRecipes';
import { Home } from './pages/Home';
import { Navbar } from './components/Navbar';
import './css/index.css';
import { Explore } from './pages/Explore';
import { ShoppingCart } from './pages/ShoppingCart';

function App() {
  return (
    <div>
      <Navbar />
      {/* @ts-ignore*/}
      <Routes basename="/">
        <Route path="/" element={<Home />} />
        <Route path="/recipe/:id" element={<DisplayOne />} />
        <Route path="/edit/:id" element={<EditRecipes />} />
        <Route path="/new" element={<CreateNew />} />
        <Route path="/explore" element={<Explore />} />
        <Route path="/cart" element={<ShoppingCart />} />
      </Routes>
    </div>
  );
}

export default App;
