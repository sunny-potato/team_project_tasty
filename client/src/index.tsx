import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './App';
import { HashRouter, BrowserRouter } from 'react-router-dom';

ReactDOM.render(
  /* Would prefer to change to server solution if avail instead of hashrouter */
  <HashRouter>
    <App />
  </HashRouter>,
  document.getElementById('root')
);
