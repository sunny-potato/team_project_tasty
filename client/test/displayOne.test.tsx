import * as React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { DisplayOne } from '../src/pages/DisplayOne';

import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { fireEvent } from '@testing-library/react';

import { Ingredient } from '../src/DataService';

let container: any = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('DisplayOne page tests', () => {
  test('DisplayOne page draws correctly (snapshot) including data from test recipe 1', async () => {
    await act(async () => {
      await render(
        <MemoryRouter initialEntries={['/recipe/1']}>
          <Routes>
            <Route path="/recipe/:id" element={<DisplayOne />} />
          </Routes>
        </MemoryRouter>,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });

  test('Changing portions', async () => {
    await act(async () => {
      await render(
        <MemoryRouter initialEntries={['/recipe/1']}>
          <Routes>
            <Route path="/recipe/:id" element={<DisplayOne />} />
          </Routes>
        </MemoryRouter>,
        container
      );
    });

    // get hold of the current amount on ingredient 1
    let amount1 = container.querySelector('div table td')?.innerHTML;
    let indexOfAND = amount1.indexOf('&');
    amount1 = amount1.substring(0, indexOfAND);
    expect(amount1).toBe('1.5');

    // get hold of selector
    const selector = container.querySelector('div select');

    // change selector to 6 portions (firing event using react test library)
    fireEvent.change(selector, { target: { value: 6 } });

    // update and check new amount on ingredient 1
    amount1 = container.querySelector('div table td')?.innerHTML;
    indexOfAND = amount1.indexOf('&');
    amount1 = amount1.substring(0, indexOfAND);
    expect(amount1).toBe('2.3');
  });

  test('Like recipe', async () => {
    await act(async () => {
      await render(
        <MemoryRouter initialEntries={['/recipe/1']}>
          <Routes>
            <Route path="/recipe/:id" element={<DisplayOne />} />
          </Routes>
        </MemoryRouter>,
        container
      );
    });

    // confirm that this is a popular recipe
    let popular = container.querySelector('div h6.Recipe-popular');
    expect(popular.innerHTML).toBe('This item is popular');

    // click on popular icon (firing event using react test library)
    fireEvent.click(container.querySelector("div button[title='Like the recipe']"));

    // confirm that recipe is no longer popular
    popular = container.querySelector('div h6.Recipe-popular');
    expect(popular.innerHTML).toBe(' ');
  });

  test('Add to cart', async () => {
    await act(async () => {
      await render(
        <MemoryRouter initialEntries={['/recipe/1']}>
          <Routes>
            <Route path="/recipe/:id" element={<DisplayOne />} />
          </Routes>
        </MemoryRouter>,
        container
      );
    });

    // clear local storage (mocked) and check number of ingredients
    localStorage.clear();
    let cart: Ingredient[] = JSON.parse(localStorage.getItem('cart') as string);
    expect(cart.length).toEqual(0);

    // add ingredients to cart
    fireEvent.click(container.querySelector("button[title='Add recipe to shopping list']"));

    // check number of ingrediets again to verify addToCart function
    cart = JSON.parse(localStorage.getItem('cart') as string);
    expect(cart.length).toEqual(7);
  });
});
