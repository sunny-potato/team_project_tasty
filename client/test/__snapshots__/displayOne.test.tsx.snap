// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayOne page tests DisplayOne page draws correctly (snapshot) including data from test recipe 1 1`] = `
<div>
  <div
    class="Content-main"
  >
    <h1
      class="Recipe-title"
    >
      One-Pot Mac and Cheese
    </h1>
    <h5
      class="Recipe-tags"
    >
      Tags: 
      Dinner
    </h5>
    <h6
      class="Recipe-new"
    >
      This is a new recipe!
    </h6>
    <h6
      class="Recipe-popular"
    >
       
    </h6>
    <svg
      class="Icon-print"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Print the shopping list
      </title>
      <path
        d="M2.5 8a.5.5 0 1 0 0-1 .5.5 0 0 0 0 1z"
      />
      <path
        d="M5 1a2 2 0 0 0-2 2v2H2a2 2 0 0 0-2 2v3a2 2 0 0 0 2 2h1v1a2 2 0 0 0 2 2h6a2 2 0 0 0 2-2v-1h1a2 2 0 0 0 2-2V7a2 2 0 0 0-2-2h-1V3a2 2 0 0 0-2-2H5zM4 3a1 1 0 0 1 1-1h6a1 1 0 0 1 1 1v2H4V3zm1 5a2 2 0 0 0-2 2v1H2a1 1 0 0 1-1-1V7a1 1 0 0 1 1-1h12a1 1 0 0 1 1 1v3a1 1 0 0 1-1 1h-1v-1a2 2 0 0 0-2-2H5zm7 2v3a1 1 0 0 1-1 1H5a1 1 0 0 1-1-1v-3a1 1 0 0 1 1-1h6a1 1 0 0 1 1 1z"
      />
    </svg>
    <p
      class="Recipe-description>"
    >
      Who likes cleaning up after making mac and cheese? 
          Not this girl. This one-pot mac and cheese is a family favorite, 
          and my 3-year-old is thrilled to see it coming to the dinner table. 
          We love to add sliced smoked sausage to this creamy mac recipe!
    </p>
    <div>
      <div>
        <label>
          Number of portions: 
        </label>
        <select
          name="portions"
          title="Number of portions"
        >
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
          <option
            value="9"
          >
            9
          </option>
          <option
            value="10"
          >
            10
          </option>
        </select>
      </div>
      <table
        class="Ingredients-main"
      >
        <tbody>
          <tr>
            <th
              class="Ingredients-header"
            >
              Ingredients:
            </th>
            <th />
          </tr>
          <tr
            class="Ingredients-row"
          >
            <td
              class="Ingredients-cell"
            >
              1.5
               
              cup
            </td>
            <td
              class="Ingredients-name"
            >
              whole milk
            </td>
          </tr>
          <tr
            class="Ingredients-row"
          >
            <td
              class="Ingredients-cell"
            >
              3
               
              cup
            </td>
            <td
              class="Ingredients-name"
            >
              water
            </td>
          </tr>
          <tr
            class="Ingredients-row"
          >
            <td
              class="Ingredients-cell"
            >
              1
               
              package
            </td>
            <td
              class="Ingredients-name"
            >
              elbow macaroni (16 ounces)
            </td>
          </tr>
          <tr
            class="Ingredients-row"
          >
            <td
              class="Ingredients-cell"
            >
              4
               
              ounce
            </td>
            <td
              class="Ingredients-name"
            >
              velveeta, cubed
            </td>
          </tr>
          <tr
            class="Ingredients-row"
          >
            <td
              class="Ingredients-cell"
            >
              2
               
              cup
            </td>
            <td
              class="Ingredients-name"
            >
              shredded sharp cheddar cheese
            </td>
          </tr>
          <tr
            class="Ingredients-row"
          >
            <td
              class="Ingredients-cell"
            >
              0.5
               
              teaspoon
            </td>
            <td
              class="Ingredients-name"
            >
              salt
            </td>
          </tr>
          <tr
            class="Ingredients-row"
          >
            <td
              class="Ingredients-cell"
            >
              0.5
               
              teaspoon
            </td>
            <td
              class="Ingredients-name"
            >
              coarsely ground pepper
            </td>
          </tr>
        </tbody>
      </table>
      <button
        class="Button-navigation"
        title="Like the recipe"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.864.046C7.908-.193 7.02.53 6.956 1.466c-.072 1.051-.23 2.016-.428 2.59-.125.36-.479 1.013-1.04 1.639-.557.623-1.282 1.178-2.131 1.41C2.685 7.288 2 7.87 2 8.72v4.001c0 .845.682 1.464 1.448 1.545 1.07.114 1.564.415 2.068.723l.048.03c.272.165.578.348.97.484.397.136.861.217 1.466.217h3.5c.937 0 1.599-.477 1.934-1.064a1.86 1.86 0 0 0 .254-.912c0-.152-.023-.312-.077-.464.201-.263.38-.578.488-.901.11-.33.172-.762.004-1.149.069-.13.12-.269.159-.403.077-.27.113-.568.113-.857 0-.288-.036-.585-.113-.856a2.144 2.144 0 0 0-.138-.362 1.9 1.9 0 0 0 .234-1.734c-.206-.592-.682-1.1-1.2-1.272-.847-.282-1.803-.276-2.516-.211a9.84 9.84 0 0 0-.443.05 9.365 9.365 0 0 0-.062-4.509A1.38 1.38 0 0 0 9.125.111L8.864.046zM11.5 14.721H8c-.51 0-.863-.069-1.14-.164-.281-.097-.506-.228-.776-.393l-.04-.024c-.555-.339-1.198-.731-2.49-.868-.333-.036-.554-.29-.554-.55V8.72c0-.254.226-.543.62-.65 1.095-.3 1.977-.996 2.614-1.708.635-.71 1.064-1.475 1.238-1.978.243-.7.407-1.768.482-2.85.025-.362.36-.594.667-.518l.262.066c.16.04.258.143.288.255a8.34 8.34 0 0 1-.145 4.725.5.5 0 0 0 .595.644l.003-.001.014-.003.058-.014a8.908 8.908 0 0 1 1.036-.157c.663-.06 1.457-.054 2.11.164.175.058.45.3.57.65.107.308.087.67-.266 1.022l-.353.353.353.354c.043.043.105.141.154.315.048.167.075.37.075.581 0 .212-.027.414-.075.582-.05.174-.111.272-.154.315l-.353.353.353.354c.047.047.109.177.005.488a2.224 2.224 0 0 1-.505.805l-.353.353.353.354c.006.005.041.05.041.17a.866.866 0 0 1-.121.416c-.165.288-.503.56-1.066.56z"
          />
        </svg>
      </button>
      <a
        href="/edit/1"
      >
        <button
          class="Button-navigation"
          title="Edit recipe"
        >
          Edit recipe
        </button>
      </a>
      <a
        href="/cart"
      >
        <button
          class="Button-navigation"
          title="Add recipe to shopping list"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M4 6.414L.757 3.172l1.415-1.415L5.414 5h15.242a1 1 0 0 1 .958 1.287l-2.4 8a1 1 0 0 1-.958.713H6v2h11v2H5a1 1 0 0 1-1-1V6.414zM6 7v6h11.512l1.8-6H6zm-.5 16a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm12 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"
              />
            </g>
          </svg>
           
        </button>
      </a>
      <div>
        <a
          href="/"
        >
          <button
            class="Button-navigation"
            title="Back"
          >
            Back
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;
